README.md

```diff
@@ -25,16 +25,44 @@ Once you have implemented the `sa` program (and tested it to the best of your ab
 
 ## Algorithm
 
-*Which algorithm did you use to construct the suffix array. What is its running time?*
+In our project we have implemented the O(n²) algorithm that uses radix sort instead of the naive comparison based sorting.
+Bucket sort can be done in time O(n+k), where n is the length of our array (strings of length n have n or n+1 suffixes). The key is our alphabet, which for numbers would be digits from 0-9. In our case the alphabet is unknown but we have taken the liberty to assume that it is no more than a bytesize (256), since most normal characters fits in this range. This means k is constant and we can describe Bucket sort as O(n). The longest string we sort on has length n and therefore we need to do Bucket sort n times, resulting in the running time O(n²).
 
+For the search part we have implemented a binary search function that finds upper bounds and lower bounds. Our variant of binary search runs O(m * log(n)). This is the running time for both finding lower and upper bound. We start with array size n and divide search space by two each iteration, thus a total of O(log n) iterations. In each iteration we compare the pattern m, taking up to O(m).  
+This of course only the time it takes to return two indexes. The time it would take to get the actual results in SAM or similar would obviously be bounded by z.
 ## Insights you may have had while implementing the algorithm
 
+One of the most interesting insights was the great difference in running time for construction vs searching.
+This also made it a bit complicated to validate the searching part, since long search times were hard to construct.
+
 ## Problems encountered if any
 
+We had some troubles identifiying the different edge cases for the binary search algorithm more specifically the upper and lower bound algorithms. 
```

Comment:
Whenever you find it difficult to work out cases, you should write down the invariant you are using. For finding lower or upper bounds, there shouldn't be _any_ special cases.

Consider the lower bound. That is the index at which values in the array go from below the value of interest, call it $v$, to above or equal $v$. There technically _is_ a special case because we have to define negative indices as being below $v$ if we want to get index zero when all elements are greater, but that won't come into play for the algorithm.

You have a switch point (the lower bound) between indices $l$ and $h$ (low and high) and you have the invariant that all indices _below_ $l$ is smaller than $v$ and all values above $h$ are greater than or equal to $v$.

![lower-bound](https://user-images.githubusercontent.com/3206538/200117399-ff0077ee-5d37-40f6-9619-4106c31aea14.png)

If $l=h$ then $l$ must necessarily be the switch point, because the invariant tells us that every point below $l$ is smaller than $v$, so that is a termination condition where we have found what we are searching for.

If you have a point, $m$, between $l$ and $h$, $l\leq m < h$ (picking the mid-point gives the search logarithmic time, but any point will do), then we can reduce the interval.
- If $x[m] < v$ we can move $l$ up to $m+1$ (it is safe to move it higher than $m$ because $x[m] < v$, and since we allow $l=m$ we must move it higher to ensure termination).
- if $x[m] \geq v$ we can move $h$ down to $m$. That still satisfies the invariant that $x[i]\geq v$ for all $i\geq h$, and because $h$ is always larger than $m$ we know that we reduce the interval, taking us closer to termination.

There aren't any further cases.

```go
	for lo < hi {
		m := (lo + hi) / 2
		if x[m] < a {
			lo = m + 1
		} else {
			hi = m
		}
	}
```

Of course, if you also have to index into the strings at some offset, there _are_ more cases, but those are mostly related to avoiding out-of-bounds access. Write a function for getting the index in a file. It could give you zero when you are beyond the lenght, or you can index modulo the string length. Either will work.

```go
	for lo < hi {
		m := (lo + hi) / 2
		if safeIdx(x, sa[m]+offset) < a {
			lo = m + 1
		} else {
			hi = m
		}
	}
```

If you work with discrete values, an easy trick to get the upper bound as well is to notice that `lowerBound(x,v+1) == upperBound(x,v)`. This doesn't work as easily with strings, though, because a string plus one isn't that well defined (you can define it by appending the smallest letter, but then it is expensive, so ignore that). Then you need to slightly modify the invariant and go from there.